### ISSUE TITLE
docker-compose 1.27.3 config depends_on breaks stack deploy

### ISSUE THREAD

**Issue by dfresh613** [docker/compose#7773]
## Description of the issue
When we apply docker-compose config to a docker-compose file with depends_on, the end result is not deployable by docker stack..

## Steps to reproduce the issue
Example docker-compose file:
```
services:
  test:
    depends_on:
    - test2
    image: a/fakeimage:latest
  test2:
    image: a/fakeimage:latest
version: '3.4'
```

`Use docker-compose -f <example-config> config > new_config`

The result is:

```
services:
  test:
    depends_on:
      test2:
        condition: service_started
    image: a/fakeimage:latest
  test2:
    image: a/fakeimage:latest
version: '3.4'

```

Now, attempting to deploy that new compose config, docker stack deploy fails:
```
docker stack deploy -c configout.yml iservices
services.test.depends_on must be a list
```

This behavior does not occur in 1.26.x or 1.27.2

docker-ce == 19.03.5
docker-compose ==1.27.3

**Comment by ulyssessouza** [comment]
Hello @dfresh613.

Actually the new output is compliant with the new `compose-spec`.
Could you please explain your usecase of using the output of `docker-compose config` as input for `docker stack deploy`?

**Comment by dfresh613** [comment]
We have a docker-compose file to deploy an application that may be deployed either as a stack, or with docker-compose, easiest to maintain one config file for both, and consistently apply updates using other compose files (docker-compose config -f <file1> -f <file2> config

Using docker-compose config allows us to override and apply updates to the same environment file.

Regardless our use case, this is a change that doesn't seem like it should have been made in a patch (.3) release.

**Comment by samueldeklund** [comment]
We are having a similar issue with piping `docker-compose config` output to `docker stack deploy` command.

We use several different environment files and select one based on which server we are deploying to, and use the `--env-file` parameter to inject those values into the yml file. But now that the presence of the `deploy` key throws an error instead of passing it through (even suppressing the error with `--log-level` doesn't re-enable output), our workflow is broken.

**Comment by crossdot** [comment]
I think docker-compose should respect version specified in docker-compose.yml
For 2.* it should be object map with conditions
For 3.* it should be list

**Comment by dfresh613** [comment]
What I find quite funny, and correct me if I'm wrong-- But it seems this change originally failed the existing unit tests which ensures depends_on matches type list: https://ci-next.docker.com/public/blue/organizations/jenkins/compose/detail/PR-7761/1/pipeline/#step-89-log-867

And then, in the next commit these screaming unit tests were "updated" by deleting them, or changing their validation criteria: https://github.com/docker/compose/commit/fa720787d62cf833ed7648ff55dbd248267a5b74

**Comment by SMFloris** [comment]
Indeed, we find the same issue. We have multiple compose files (one for prod, one for dev), where we deploy something like this (we make use of the interpolation of docker-compose config):

```
docker-compose -f docker-compose.prod.yml -f docker-compose.dev.yml config | docker stack deploy
```

This ensures that we run our containers as close as possible to their prod counterparts locally, the only differences being the cpu allocations, scales, labels and volumes mount.

There are multiple issues that are preventing me from even testing our stack on my laptop (i.e. #7774). I'm going to be working from my laptop in a week from now, but can't even deploy my stack since I have the newer version that came with docker. The main selling point of docker stack and docker-compose is that you can use almost the same files for both. And this has been broken so this should be treated as a regression in my opinion.

@ulyssessouza, as I saw regarding #7774, docker stack doesn't care about `compose-spec` in their latest 3.9 schema, even though in the compose-spec repository its being advertised that it has an implementation of `compose-spec`. This is untruthful. One should expect that as more and more people upgrade to Docker 3.0.5 for Mac, many more bugs like this will pop up.

Basically, I hope this was just a roadmap mismatch (i.e. the docker cli team wasn't fast enough to implement the compose spec and you guys were a bit more ambitious). This, thus, becomes a release problem. One should not release a set of tools that should be compatible with each-other that aren't and break your flow and scripts.

**Comment by SMFloris** [comment]
For everyone having issues with docker-compose config and piping the output into docker stack deploy, there is this open issue ( https://github.com/docker/cli/issues/2365 ) for which I have done a PR. Feel free to post your own use cases there in order for the Docker team to gather more data in order for them to make the right call regarding that open issue.

I did my best to make sure that the configs are merged/interpolated the same way and that the output piped back to docker stack deploy is always valid. Feel free to test that as well if you have the time.

**Comment by dfresh613** [comment]
@SMFloris that is awesome, I definitely agree that functionality is needed directly from the stack command.

That sounds feasible for a longer term solution- but I would still like to know what the planned next steps are for the current 1.27.x line of docker-compose since:

- This code changed failed unit tests, and is a regression, significantly changing expected behavior
- This backward-compatibility breaking regression was introduced in a patch release (.3), which is not best practice. 
- This oversight essentially removed supported functionality in a way that was not formally deprecated, announced, or documented, in a release version that is supposed to be stable.

I believe this should be addressed directly within the 1.27.x line. Namely an updated patch release 1.27.4 which corrects the regression and passes originally defined unit tests covering 1.27.x.

**Comment by ndeloof** [comment]
IMHO this is not a regression nor a desirable behavior for `docker-compose config` to **remove** `depends_on` set by input. This command has not been intended to act as a pre-processor for `docker stack deploy`, but to validate and render the actual compose model after applying overrides and interpolations. It's not designed to "convert a v2 file into v3", which actually doesn't make much sense, especially now with the compose-spec having dropped the `version` field.

**Comment by dfresh613** [comment]
The future direction makes sense, however that is not the point of the bug.

The point is that this **_backward compatibility breaking change_** (as evidenced by the existing unit tests which were deleted) was improperly added in a patch release.

Per [standard semantic versioning ](https://semver.org/) if my package requirement is `docker-compose~=1.27.0` I would expect docker-compose 1.27.x to not introduce a  compatibility breaking bug that would cause my already in production automation to break (using ~=1.27.0).

If you want to deprecate functionality and no longer support things as they once were, fine. But please remember to do so in a standards based (semantic versioning) fashion, preferably with some sort of deprecation notice.

Failing to do so makes docker-compose less of an appealing option to use in real production environments.

**Comment by ndeloof** [comment]
semantic versioning expectatons are actually incredibly hard to follow on real world project, as any change can, intentionally or not, introduce sort of regression. I only can imagine this as a "best effort" approach. And, again, you're using `config` in a way it was designed for, the fact it silently removed `depends_on` could be considered a bug that we fixed. If you rely on a buggy behaviour, and we have to adopt strict semantic versioning, then each new release should be a major release :P

Bythe way, as you understand the future direction, better embrace it and don't focus on what's left behind. Life's short

**Comment by dfresh613** [comment]
``` and we have to adopt strict semantic versioning, then each new release should be a major release ```

If your code maintenance and testing practices are that unreliable, then yes it should be a major release every time.


So to summarize - docker-compose does not follow semantic version standards, and anyone using it in production should upgrade at their own risk?

**Comment by ndeloof** [comment]
This is not what I'm saying. docker-compose include a significant test suite to catch regression and the release process is based on release candidates to ensure we don't have bad impact on the ecosystem.
What I'm saying is that, whatever we do, there might be some regression we didn't captured first. Any software upgrade is one way or the other a "risk".

And, again: what's has been described here as a "regression" could be considered a bug we fixed. Fixing bugs doesn't require a major release. `docker-compose config` never claim to make a compose file compliant with `docker stack` by dropping unsupported attributes. We can't support "features" that are just unsupport usages.

![image](https://user-images.githubusercontent.com/132757/106021538-30e3ee80-60c5-11eb-807e-ab5391b96f40.png)

**Comment by evictor** [comment]
Here's a drop-in workaround that employs [`yq`](https://github.com/mikefarah/yq), the Yaml query lib inspired by the JSON equivalent [`jq`](https://stedolan.github.io/jq/). The output from `docker-compose config` will have all `service.*.depends_on` mapped to array-of-string service names, the compatible type for `docker stack deploy`. Enjoy. :)

```shell
docker-compose config \
    | yq e '(.services[] | select(.depends_on | tag == "!!map")).depends_on |= (. | keys)' -

# .services[] - for each service map
# select(.depends_on | tag == "!!map") - keep service if depends_on is a map
#     (discards services where depends_on is null/undefined or already string[])
# (...).depends_on |= (...) - assign to the selected services' depends_on with
#     the depends_on as context for RHS
# . | keys - map depends_on (as a map of service names => condition) to list
#     of map keys (service names)
```

I'm glad @SMFloris put up a PR with something forward thinking but IMO, it's a shame someone on the team wouldn't put forth what would be just a little effort to adopt what was undefined behavior into the product to make it defined (the interaction between `docker-compose config` and `docker stack deploy`). Why the `config` command could not simply be included under `docker stack` with appropriate changes is beyond me. You get a bunch of wins:

1. The features (env subbing, etc.) from `docker-compose config` are now suddenly first-class features of Swarm CLI
2. Minimal new code to be written—probably no new code written, just some lines removed or tweaked slightly to make sure output is valid for Swarm
3. Two separate versions of `config` subcommand seems reasonable if the two are expected to drift apart. If they're in 2 separate codebases, great. If they're in 1 codebase, then include/import/abstract your way to DRYness and call it a day. No sin here...

**Comment by jjakob** [comment]
This is still an issue for me. Is the "correct" generator for `docker stack deploy` `docker compose convert`?https://docs.docker.com/engine/reference/commandline/compose_convert/

**Comment by ndeloof** [comment]
There's no such thing as a "generator" for `docker stack deploy`. This command directly consume a compose file

**Comment by jjakob** [comment]
Yes, but we generate that compose file by using (currently) `docker-compose config -f docker-compose.common.yml -f docker-compose.dev.yml -f docker-compose.secrets.yml > docker-compose.yml` where this will not only merge those 3 compose files, but populate all variables from .env file and environment. Thus the product is a merged static compose file. I was asking if `docker compose convert` will provide the same functionality (merging and variable substitution from .env) and provide a merged output file suitable for use with `docker stack deploy`.

**Comment by ndeloof** [comment]
`docker compose convert` (alias `config`) will generate a canonical file for docker engine applying the compose specification. As long as `docker stack` commands are not compliant with the compose spec, there's no guarantee it will be able to parse it.

Also see https://github.com/docker/cli/pull/3257 which prevent `stack deploy` command to reject a file missing a `version` element.

**Comment by jawys** [comment]
/cc https://github.com/docker/cli/issues/3880#issuecomment-1805903513

I posted the linked comment in another repo because of the `depends_on` keyword but the discussion here seems to be very related.
